packages/web-config-server/src/tests/apiV1/dataBuilders/generic/reportServer/reportServerDataBuilder.test.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { expect } from 'chai';
+import sinon from 'sinon';
+
+import { ReportServerBuilder } from '/apiV1/dataBuilders/generic/reportServer/reportServerDataBuilder';
+
+const reportRequestKey = (reportCode, query = {}, body = {}) =>
+  `reportCode:${reportCode},${Object.entries(query)
+    .map(entry => entry.join(':'))
+    .join(',')},${Object.entries(body)
+    .map(entry => entry.join(':'))
+    .join(',')}`;
+
+const REPORT_SERVER_RESPONSES = {
+  [reportRequestKey('1', { organisationUnitCodes: 'TO', hierarchy: 'psss' })]: {
+    results: [
+      { period: '2018', organisationUnit: 'TO', dataElement: 'AGGR01', value: 1 },
+      { period: '2019', organisationUnit: 'TO', dataElement: 'AGGR02', value: 3 },
+      { period: '2020', organisationUnit: 'TO', dataElement: 'AGGR01', value: 4 },
+      { period: '2021', organisationUnit: 'TO', dataElement: 'AGGR02', value: 5 },
+    ],
+  },
+  [reportRequestKey('1', {
+    organisationUnitCodes: 'TO',
+    hierarchy: 'psss',
+    startDate: '2020-01-01',
+    endDate: '2021-01-01',
+  })]: {
+    results: [
+      { period: '2020', organisationUnit: 'TO', dataElement: 'AGGR01', value: 4 },
+      { period: '2021', organisationUnit: 'TO', dataElement: 'AGGR02', value: 5 },
+    ],
+  },
+  [reportRequestKey('2', {
+    organisationUnitCodes: 'PG',
+    hierarchy: 'strive',
+    startDate: '2020-01-01',
+  })]: {
+    results: [
+      { period: '2020', organisationUnit: 'PG', dataElement: 'PSSS01', value: 4 },
+      { period: '2021', organisationUnit: 'PG', dataElement: 'PSSS02', value: 5 },
+    ],
+  },
+};
+
+const PROJECTS = [
+  { code: 'ps', entity_hierarchy_id: '1' },
+  { code: 'str', entity_hierarchy_id: '2' },
+];
+
+const ENTITY_HIERARCHIES = [
+  { id: '1', name: 'psss' },
+  { id: '2', name: 'strive' },
+];
+
+const req = { session: { userJson: { userName: 'test' } } };
+
+const fetchReport = sinon.stub();
+fetchReport.callsFake(
+  (reportCode, query, body) => REPORT_SERVER_RESPONSES[reportRequestKey(reportCode, query, body)],
+);
+
+const findProject = sinon.stub();
+findProject.callsFake(({ code }) => PROJECTS.find(project => project.code === code));
+
+const findEntityHierarchyById = sinon.stub();
+findEntityHierarchyById.callsFake(id => ENTITY_HIERARCHIES.find(hierarchy => hierarchy.id === id));
+
+const reportConnection = { fetchReport };
+const models = {
+  project: { findOne: findProject },
+  entityHierarchy: { findById: findEntityHierarchyById },
+};
+
+describe('ReportServerDataBuilder', () => {
+  const assertBuilderResponseIsCorrect = async ({ config, query, entity }, expectedResponse) => {
+    const dataBuilder = new ReportServerBuilder(req, models, config, query, entity);
+    dataBuilder.reportConnection = reportConnection;
```

Comment:
Being quite heavy handed with the stubbing and mocking here. In truth I suspect there's a way to do this _correctly_ using sinon however I've really struggled with mocking microservices before. I'm not the most concerned when it comes to doing mocking the right way provided it works correctly, but am happy to put the time in to get this right if reviewer has a different opinion

Reply:
There are a couple of things you could change if you wanted to:
- There's no huge benefit to having data come back in the response, unless I'm missing a processing step that's required in the data builder. Instead you could just have it respond with nothing but use the spy that sinon creates to check that it was called with the right parameters. It would make things a bunch simpler, but only do it if you are interested in learning about spies (and I'm happy to talk to you more about them on a call)
- Another potential change would be to move the test data into a `.fixtures` file and/or the stubs into a `.stubs` file. That can tidy up the tests file a fair bit, though it's definitely not necessary when neither are being used across multiple tests.

However, I'm also 100% fine with the approach you've taken, and the tests themselves are good. Also, my personal philosophy is that tests should be the gentlest part of a PR review, given that push back on tests will make adding tests feel like even more work, so make it less likely that people will do them. It's a bit like my Mum's philosophy of "never criticise the way someone does the vacuuming" ;-)

Reply:
> There's no huge benefit to having data come back in the response, unless I'm missing a processing step that's required in the data builder.
You're right, and originally I just thought about spying to see if the correct URL was used. However, technically there's the conversion of `{ results: [...], ... }` to `{ data: [...] }` being performed. While totally trivial, there's actually so little going on in this databuilder that it amounts to a reasonable percentage of the logic 😂 I guess I always wanted to validate the output as it will force someone to update the tests if they decide to extend/alter the shape of the response (ie. adding 'metadata' or 'period') fields that other databuilders produce.

Reply:
> "never criticise the way someone does the vacuuming"

I love this ❤️